glean-core/src/event_database/mod.rs

```diff
@@ -0,0 +1,325 @@
+    ///
+    /// Secondly, if this is the first time the application has been run since
+    /// rebooting, any pings containing events are assembled into pings and cleared
+    /// immediately, since their timestamps won't be compatible with the timestamps
+    /// we would create during this boot of the device.
+    ///
+    /// # Arguments
+    ///
+    /// * `glean` - The Glean instance.
+    pub fn on_ready_to_send_pings(&self, glean: &Glean) {
+        match self.load_events_from_disk() {
+            Ok(_) => self.send_all_events(glean),
+            Err(err) => log::error!("Error loading pings from disk: {}", err),
+        }
+    }
+
+    fn load_events_from_disk(&self) -> Result<()> {
+        let _lock = self.file_lock.read().unwrap();
+        let mut db = self.event_stores.write().unwrap();
+        for entry in fs::read_dir(&self.path)? {
+            let entry = entry?;
+            if entry.file_type()?.is_file() {
+                let store_name = entry.file_name().into_string()?;
+                let file = BufReader::new(File::open(entry.path())?);
+                db.insert(
+                    store_name,
+                    file.lines()
+                        .filter_map(|line| line.ok())
+                        .filter_map(|line| serde_json::from_str::<RecordedEventData>(&line).ok())
+                        .collect(),
+                );
+            }
+        }
+        Ok(())
+    }
+
+    fn send_all_events(&self, glean: &Glean) {
+        let store_names = {
+            let db = self.event_stores.read().unwrap();
+            db.keys().cloned().collect::<Vec<String>>()
+        };
+
+        for store_name in store_names {
+            if let Err(err) = glean.send_ping_by_name(&store_name, false) {
+                log::error!(
+                    "Error flushing existing events to the {} ping: {}",
+                    store_name,
+                    err
+                );
+            }
+        }
+    }
+
+    /// Record an event in the desired stores.
+    ///
+    /// # Arguments
+    ///
+    /// * `glean` - The Glean instance.
+    /// * `meta` - The metadata about the event metric. Used to get the category,
+    ///   name and stores for the metric.
+    /// * `timestamp` - The timestamp of the event, in nanoseconds. Must use a
+    ///   monotonically increasing timer (this value is obtained on the
+    ///   platform-specific side).
+    /// * `extra` - Extra data values, mapping strings to strings.
+    pub fn record(
+        &self,
+        glean: &Glean,
+        meta: &CommonMetricData,
+        timestamp: u64,
+        extra: Option<HashMap<String, String>>,
+    ) {
+        let event = RecordedEventData {
+            timestamp,
+            category: meta.category.to_string(),
+            name: meta.name.to_string(),
+            extra,
+        };
+        let event_json = serde_json::to_string(&event).unwrap();
+        let mut stores_to_send: Vec<&str> = Vec::new();
+
+        {
```

Comment:
The actual storing of the event is hidden a bit in that multi-line block below.
How about we mark the different phases with a comment?
E.g. above here:

```rust
// Store the event in-memory and persist to disk
```